File: src/main/java/javaPractice/java8/lets_get_lazy/eager_vs_lazy_scala/sample.scala
def expensiveComputation(input: Int) = {
  println("called expensive computation")
  input
}

val somevalue = 4

println("Short circuit")
if(somevalue > 5 && expensiveComputation(5) > 0)
  println("")
else
  println("")
  
println("Not so efficient")
val temp = expensiveComputation(5)
if(somevalue > 5 && temp > 0)
  println("")
else
  println("")

println("But we can be quite easily...")
lazy val temp2 = expensiveComputation(5)
if(somevalue > 5 && temp2 > 0)
  println("")
else
  println("")

File: src/main/java/javaPractice/java8/lets_get_lazy/lazy_with_lambda/Sample.java
package javaPractice.java8.lets_get_lazy.lazy_with_lambda;

import java.util.*;
import java.util.function.Supplier;

public class Sample {
  public static int expensiveComputation(int input) {
    System.out.println("called expensive computation");
    return input;
  }
  
  public static void main(String[] args) {
    int somevalue = 4;
    
    System.out.println("Short circuit");
    if(somevalue > 5 && expensiveComputation(5) > 0)
      System.out.println("");
    else
      System.out.println("");
      
    System.out.println("Lazy with Lambda");
    Supplier<Integer> temp = () -> expensiveComputation(5);
    if(somevalue > 5 && temp.get() > 0)
      System.out.println("");
    else
      System.out.println("");
  }
}
